subsys/bluetooth/audio/pbp.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/types.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/audio/audio.h>
+#include <zephyr/bluetooth/audio/pbp.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/check.h>
+
+LOG_MODULE_REGISTER(bt_pbp, CONFIG_BT_PBP_LOG_LEVEL);
+
+/* PBA Service UUID + Public Broadcast Announcement features + Public_Info metadata of length 1 */
+#define MIN_PBA_SIZE			5
+
+int bt_pbp_get_announcement(const uint8_t meta[], size_t meta_len,
+			    enum bt_pbp_announcement_feature features,
+			    struct net_buf_simple *pba_data_buf)
+{
+	CHECKIF(pba_data_buf == NULL) {
+		LOG_DBG("No buffer provided for advertising data!\n");
+
+		return -EINVAL;
+	}
+
+	CHECKIF((meta == NULL && meta_len != 0) || (meta != NULL && meta_len == 0)) {
+		LOG_DBG("Invalid metadata combination: %p %zu", meta, meta_len);
+
+		return -EINVAL;
+	}
+
+	CHECKIF(pba_data_buf->size < meta_len + 3) {
+		LOG_DBG("Buffer size needs to be at least %d!\n", meta_len + 3);
+
+		return -EINVAL;
+	}
+
+	/* Fill Announcement data */
+	net_buf_simple_add_le16(pba_data_buf, BT_UUID_PBA_VAL);
+	net_buf_simple_add_u8(pba_data_buf, features);
+	net_buf_simple_add_mem(pba_data_buf, meta, meta_len);
+
+	return 0;
+}
+
+int bt_pbp_parse_announcement(struct bt_data *data,
+			      enum bt_pbp_announcement_feature *source_features,
+			      uint8_t *meta)
+{
+	size_t meta_len = 0;
+	struct bt_uuid_16 adv_uuid;
+
+	CHECKIF(!data || !source_features || !meta) {
+		return -EINVAL;
+	}
+
+	if (data->data_len < (BT_UUID_SIZE_16 + MIN_PBA_SIZE)) {
+		return -EBADMSG;
+	}
+
+	if (data->type != BT_DATA_SVC_DATA16) {
+		return -EINVAL;
+	}
+
+	if (!bt_uuid_create(&adv_uuid.uuid, data->data, BT_UUID_SIZE_16)) {
+		return -EINVAL;
+	}
+
+	if (bt_uuid_cmp(&adv_uuid.uuid, BT_UUID_PBA)) {
+		return -EBADMSG;
+	}
+
+	/* Copy source features and metadata from the Public Broadcast Announcement */
+	*source_features = data->data[BT_UUID_SIZE_16];
+	meta_len = data->data_len - (BT_UUID_SIZE_16 + sizeof(uint8_t));
+	memcpy(meta, data->data + (BT_UUID_SIZE_16 + sizeof(uint8_t)), meta_len);
```

Comment:
Isn't this missing a `sizeof(meta_len)`? 
i.e.
```suggestion
	memcpy(meta, data->data + (BT_UUID_SIZE_16 + sizeof(uint8_t) + sizeof(meta_len)), meta_len);
```

Reply:
Shouldn't the metadata also include the meta length? From what I understand from the BT_AUDIO_CODEC_DATA macro the structure we use is:  meta_len, meta_type, meta.